driver/test/src/main/java/org/opennms/oce/driver/test/TestDriver.java

```diff
@@ -60,10 +61,55 @@ public TestDriver(EngineFactory engineFactory) {
         return run(alarms, Collections.emptyList());
     }
 
+    /**
+     * Drives a Correlation Engine and tests service resumption by bifurcating the 
+     *  Alarms into a before and after at <code>timestamp</code>.
+     *  
+     *  Alarms before the timestamp will be reduced so that only the latest Alarm with a given Id will be included.
+     *  Additionally, if that alarm is a CLEAR, no Alarm with ahat Id will be passed to the Engine.
+     *  This approximates what would be available to an engine on a restart.
+     * 
+     * @param alarms The alarms to replay
+     * @param inventory The inventory to use
+     * @param timestamp the Driver emulates an engine reset at timestamp
+     * @return
+     */
+    public List<Incident> run(List<Alarm> alarms, List<InventoryObject> inventory, long timestamp) {
+        // bifurcate the alarms
+        final List<Alarm> sortedAlarms = alarms.stream()
+                .sorted(Comparator.comparing(Alarm::getTime).thenComparing(Alarm::getId))
+                .collect(Collectors.toList());
+        List<Alarm> previousAlarms = reduceAlarms(sortedAlarms.stream().filter(a -> a.getTime() <= timestamp).collect(Collectors.toList()));
+        List<Alarm> afterAlarms = sortedAlarms.stream().filter(a -> a.getTime() > timestamp).collect(Collectors.toList());
+        // Run the engine 
+        List<Incident> previousIncidents = run(previousAlarms, inventory);
+        List<Incident> allIncidents = new ArrayList<>(previousIncidents);
+        // re-run the engine 
+        allIncidents.addAll(run(afterAlarms, inventory, previousAlarms, previousIncidents));
+        return allIncidents;
+    }
+
+    /**
+     * Assumes a sorted list of alarms.
+     * @param in A time sorted List of Alarms
+     * @return The last Alarm of each Id with No Alarm for a given Id if the last Alarm for that Id had a Severity of CLEAR. 
+     */
+    static List<Alarm> reduceAlarms(List<Alarm> in) {
+        Map<String, Alarm> reduced = new HashMap<>();
+        for (Alarm a : in) {
+            reduced.put(a.getId(), a);
+        }
+        return reduced.values().stream().filter(a -> !a.isClear()).collect(Collectors.toList());
```

Comment:
Not sure if it matters here or not, but the resulting list may no longer be sorted.